pkg/pxc/statefulset.go

```diff
@@ -132,3 +134,45 @@ func getConfigVolumes(cr *api.PerconaXtraDBCluster, cvName string) corev1.Volume
 	vol1.ConfigMap.Optional = &t
 	return vol1
 }
+
+func getDataVolumes(cr *api.PerconaXtraDBCluster, sfs api.StatefulApp) (isPVC bool, pvcs []corev1.PersistentVolumeClaim, dataVolume corev1.VolumeSource) {
+	// 1. check whether pvc is existed.
+	if cr.Spec.PXC.VolumeSpec.PersistentVolumeClaim != nil {
+		isPVC = true
+		pvcs = sfs.PVCs(&cr.Spec.PXC.VolumeSpec)
+		return
+	}
+
+	// 2. check whether hostPath is existed.
+	if cr.Spec.PXC.VolumeSpec.HostPath != nil {
+		dataVolume.HostPath = cr.Spec.PXC.VolumeSpec.HostPath
+		return
+	}
+
+	// 2. check whether emptyDir is existed.
+	if cr.Spec.PXC.VolumeSpec.EmptyDir != nil {
+		dataVolume.EmptyDir = cr.Spec.PXC.VolumeSpec.EmptyDir
+		return
+	}
+
+	return
+}
+
+func getVolumes(cr *api.PerconaXtraDBCluster, sfs api.StatefulApp, cvName string) (bool, []corev1.PersistentVolumeClaim, []corev1.Volume) {
+	isPVC, pvcs, dataVolume := getDataVolumes(cr, sfs)
+	volumes := []corev1.Volume{
+		corev1.Volume{
+			VolumeSource: dataVolume,
+		},
+	}
+
+	switch sfs.(type) {
+	case *statefulset.Node:
+		configMapVolume := getConfigVolumes(cr, cvName)
+		volumes[0].Name = dataVolumeName
+		volumes = append(volumes, configMapVolume)
+	case *statefulset.Proxy:
+		volumes[0].Name = proxydataVolumeName
+	}
+	return isPVC, pvcs, volumes
+}
```

Comment:
Let's rearrange and/or extend `v1alpha1.App` interface instead of all of this "ifs" and types casting.